modyn/model_storage/internal/model_storage_manager.py

```diff
@@ -0,0 +1,280 @@
+
+        Args:
+            pipeline_id: the pipeline identifier for the model.
+            trigger_id: the trigger associated with the model.
+            checkpoint_path: path to the checkpoint containing the model.
+
+        Returns:
+            int: the model id which identifies the stored model.
+        """
+        checkpoint = torch.load(checkpoint_path)
+        policy = self.get_model_storage_policy(pipeline_id)
+
+        # split the model (stored under the "model" key) from metadata.
+        assert "model" in checkpoint
+        state_dict = checkpoint["model"]
+        local_model_filename = f"{current_time_millis()}_{pipeline_id}_{trigger_id}.model"
+        model_path = self._storage_dir / local_model_filename
+
+        # handle the new model according to the model storage policy. If it is stored incrementally, we receive
+        # the model id of the parent.
+        parent_id = self._handle_new_model(pipeline_id, trigger_id, state_dict, model_path, policy)
+        checkpoint.pop("model")
+
+        # now checkpoint only contains optimizer state and metadata.
+        local_metadata_filename = f"{current_time_millis()}_{pipeline_id}_{trigger_id}.metadata.zip"
+        metadata_path = self._storage_dir / local_metadata_filename
+        torch.save(checkpoint, metadata_path)
+
+        # add the new model to the database.
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            return database.add_trained_model(
+                pipeline_id, trigger_id, local_model_filename, local_metadata_filename, parent_id
+            )
+
+    def _handle_new_model(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        state_dict: dict,
+        model_path: pathlib.Path,
+        policy: ModelStoragePolicy,
+    ) -> Optional[int]:
+        """
+        Handle the new model according to the model storage policy.
+
+        Args:
+            pipeline_id: the pipeline, to which the model belongs.
+            trigger_id: the trigger identifier associated with the model.
+            state_dict: the model's state.
+            model_path: path, under which the model must be stored.
+            policy: the model storage policy applied to store the model.
+        Returns:
+            int: if the model is stored incrementally, the parent model id is returned.
+        """
+
+        # check whether we must apply the incremental storage strategy or the full model strategy.
+        if policy.incremental_model_strategy and (
+            policy.full_model_interval is None or trigger_id % policy.full_model_interval != 0
+        ):
+            parent_model_id: Optional[int] = self._get_parent_model_id(pipeline_id, trigger_id)
+            if parent_model_id is not None:
+                # store the model according to the incremental model strategy.
+                parent_model_state = self._get_base_model_state(pipeline_id)
```

Comment:
Could we, instead of loading this here, instantiate the base model state in `reconstruct_model` in the base case? Or is there any reason to do this here? I think this would lead to less confusion on the flow, since where we actually need this is in the base case when we load the full model, right?

Reply:
No, you are right! Thanks for the catch. This simplified the workflow.